### PR TITLE
[Harness] Refactor, simplify and fix issues in the result parsing.

### PR DESCRIPTION
**Pull request by mandel-macaque**
Refactor result parsing and test it. Make sure that all the possible
corner cases are taken into account. The changes are as follow:

1. Move all the logic to a new class, which will take care of the reustl
parsing.
2. Simplify the parsing method by splitting the noise code to other
functions. There is no need to parse several files in the same method.
3. Do not create crash reports on Xml if the tests did not crash. This is the main reason we had a mismatch between harness and vsts. We consumed the crash logs and created reports even when the app did not crash, giving false negatives.
4. Add tests that make sure that the correct result is returned when:
  a. The app cannot launch.
  b. The app timesout.
  c. The app crashes.

### CODE REVIEW

File: tests/xharness/TestReporter.cs [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+					try {
+						runner.Logs.Add (crashLog);
+
+						if (pid == -1) {
+							// Find the pid
+							pid = await GetPidFromMainLog ();
+						}
+
+						GetCrashReason (pid, crashLog, out crashReason);
+						if (crashReason != null)
```
Comment (mandel-macaque) [inline tests/xharness/TestReporter.cs]: @spouliot in this line is where we have the fix for the false positives. We used to create a crash report, but we are going through the reports even when we did not crash (adding them is not bad practice). But doing the crash xml failure meant a diff failure count. Everything got moved to GenerateXmlFailures which takes a param stating if there was a crash or not.

File: tests/xharness/AppRunner.cs [hunk 2]
```diff
@@ -654,174 +442,9 @@ public async Task<int> RunAsync ()
...
-						$"App Timeout {AppInformation.AppName} {variation} on bot {deviceName}",
-						$"{AppInformation.AppName} {variation} Test run timed out after {timeout.TotalMinutes} minute(s) on bot {deviceName}.",
-						MainLog.FullPath,
-						harness.XmlJargon);
-				}
-			}
-
-			return success.Value ? 0 : 1;
+			// check the final status, copy all the required data
+			(Result, FailureMessage) = await testResult.ParseResult ();
```
Comment (mandel-macaque) [inline tests/xharness/AppRunner.cs]: Much simpler, hide all the complicated logic. We set mlaunch, execute and let the class parse the results.

File: tests/xharness/TestReporter.cs [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+						int.TryParse (line.Substring (idx, next_idx - idx), out pid);
+				}
+				if (pid != -1)
+					return pid;
+			}
+			return pid;
+		}
+
+		// return the reason for a crash found in a log
+		void GetCrashReason (int pid, ILog crashLog, out string crashReason)
```
Comment (mandel-macaque) [inline tests/xharness/TestReporter.cs]: Same code as we used to have inside the RunAsync method, just moved it out so that is easier to understand what is happening.

File: tests/xharness/TestReporter.cs [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+			doc.Load (reader);
+			foreach (XmlNode node in doc.SelectNodes ($"/root/processes/item[pid = '" + pid + "']")) {
+				Console.WriteLine (node?.InnerXml);
+				Console.WriteLine (node?.SelectSingleNode ("reason")?.InnerText);
+				crashReason = node?.SelectSingleNode ("reason")?.InnerText;
+			}
+		}
+
+		// return if the tcp connection with the device failed
+		async Task<bool> TcpConnectionFailed ()
```
Comment (mandel-macaque) [inline tests/xharness/TestReporter.cs]: Same logic, just moved it out.

File: tests/xharness/TestReporter.cs [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+					} else if (line.Contains ("error MT1008")) {
+						pidData.launchFailure = true;
+					}
+				}
+			}
+			return pidData;
+		}
+
+		// parse the main log to get the pid 
+		async Task<int> GetPidFromMainLog ()
```
Comment (mandel-macaque) [inline tests/xharness/TestReporter.cs]: Same logic, just moved it out.

File: tests/xharness/TestReporter.cs [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+			CallbackLog = new CallbackLog ((line) => {
+				// MT1111: Application launched successfully, but it's not possible to wait for the app to exit as requested because it's not possible to detect app termination when launching using gdbserver
+				waitedForExit &= line?.Contains ("MT1111: ") != true;
+				if (line?.Contains ("error MT1007") == true)
+					launchFailure = true;
+			});
+		}
+
+		// parse the run log and decide if we managed to start the process or not
+		async Task<(int pid, bool launchFailure)> GetPidFromRunLog () {
```
Comment (mandel-macaque) [inline tests/xharness/TestReporter.cs]: Same logic, just moved it out.

File: tests/xharness/TestReporter.cs [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+				timedout = true;
+			}
+		}
+
+		public async Task CollectSimulatorResult (Task<ProcessExecutionResult> processExecution)
+		{
+			isSimulatorTest = true;
+			await CollectResult (processExecution);
+
+			if (!Success.Value) {
```
Comment (mandel-macaque) [inline tests/xharness/TestReporter.cs]: The code is very similar between Sim and Device, we had code duplication that has been moved to CollectResult, only diff between sim and device is killing the process.

File: tests/xharness/TestReporter.cs [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+			} else if (result.Succeeded) {
+				runner.MainLog.WriteLine (completionMessage);
+				Success = true;
+			} else {
+				runner.MainLog.WriteLine (failureMessage);
+				Success = false;
+			}
+		}
+
+		public void LaunchCallback (Task<bool> launchResult)
```
Comment (mandel-macaque) [inline tests/xharness/TestReporter.cs]: Moved the logic from AppRunner here, since it is more important for the TestReporter.

File: tests/xharness/TestReporter.cs [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+
+		public async Task CollectDeviceResult (Task<ProcessExecutionResult> processExecution)
+		{
+			isSimulatorTest = false;
+			await CollectResult (processExecution);
+		}
+
+		async Task<(string ResultLine, bool Failed)> GetResultLine (string logPath)
+		{
+			string resultLine = null;
```
Comment (mandel-macaque) [inline tests/xharness/TestReporter.cs]: Same logic we had, just got moved to a method and made async since we could do it without blocking.

File: tests/xharness/TestReporter.cs [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+				}
+
+			}
+			// delete not needed copy
+			File.Delete (path);
+
+			// not the most efficient way but this just happens when we run
+			// the tests locally and we usually do not run all tests, we are
+			// more interested to be efficent on the bots
+			(parseResult.resultLine, parseResult.failed) = await GetResultLine (test_log_path);
```
Comment (mandel-macaque) [inline tests/xharness/TestReporter.cs]: The entire method is the same, but added the GetResultLine to make it cleaner.

File: tests/xharness/TestReporter.cs [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+			} else {
+				WrenchLog.WriteLine ("AddSummary: <b><i>{0} crashed</i></b><br/>", runner.RunMode);
+				runner.MainLog.WriteLine ("Test run crashed");
+				return (false, true);
+			}
+		}
+
+		// generate all the xml failures that will help the integration with the CI and return the failure reason
+		async Task GenerateXmlFailures (string failureMessage, bool crashed, string crashReason)
+		{
```
Comment (mandel-macaque) [inline tests/xharness/TestReporter.cs]: Big change, we do not longer create a crash xml failure per crash report. we only do it when we crash.

File: tests/xharness/Xharness.Tests/Tests/AppRunnerTests.cs [hunk 12]
```diff
@@ -493,10 +516,6 @@ public async Task RunOnSimulatorSuccessfullyTest ()
 			// Verify
 			Assert.AreEqual (0, result);
 
-			mainLog.Verify (x => x.WriteLine ("Test run started"));
-			mainLog.Verify (x => x.WriteLine ("Test run completed"));
-			mainLog.Verify (x => x.WriteLine ("Test run succeeded"));
```
Comment (mandel-macaque) [inline tests/xharness/Xharness.Tests/Tests/AppRunnerTests.cs]: Verifications moved to the crash reporter tests.

File: tests/xharness/TestReporter.cs [hunk 13]
```diff
@@ -0,0 +1,490 @@
...
+			runner = appRunner ?? throw new ArgumentNullException (nameof (appRunner));
+			deviceName = device; // can be null on simulators 
+			listener = simpleListener ?? throw new ArgumentNullException (nameof (simpleListener));
+			runLog = log ?? throw new ArgumentNullException (nameof (log));
+			crashLogs = new Logs (runner.Logs.Directory);
+			crashReporter = crashReports ?? throw new ArgumentNullException (nameof (crashReports));
+			Timeout = runner.GetNewTimeout ();
+			TimeoutWatch  = new Stopwatch ();
+			CallbackLog = new CallbackLog ((line) => {
```
Comment (mandel-macaque) [inline tests/xharness/TestReporter.cs]: Same logic as the one found in RunAsync. Moved here since we are interested of the wait in this class.

File: tests/xharness/Xharness.Tests/Tests/AppRunnerTests.cs [hunk 14]
```diff
@@ -650,10 +677,6 @@ public async Task RunOnDeviceSuccessfullyTest ()
 
 			processManager.VerifyAll ();
 
-			mainLog.Verify (x => x.WriteLine ("Test run started"));
-			mainLog.Verify (x => x.WriteLine ("Test run completed"));
-			mainLog.Verify (x => x.WriteLine ("Test run succeeded"));
```
Comment (mandel-macaque) [inline tests/xharness/Xharness.Tests/Tests/AppRunnerTests.cs]: Verifications added in the TestReporter tests.

File: tests/xharness/Xharness.Tests/Tests/AppRunnerTests.cs [hunk 15]
```diff
@@ -759,10 +792,6 @@ public async Task RunOnDeviceWithFailedTestsTest ()
 
 			processManager.VerifyAll ();
 
-			mainLog.Verify (x => x.WriteLine ("Test run started"));
-			mainLog.Verify (x => x.WriteLine ("Test run completed"));
-			mainLog.Verify (x => x.WriteLine ("Test run failed"));
-
```
Comment (mandel-macaque) [inline tests/xharness/Xharness.Tests/Tests/AppRunnerTests.cs]: Same, we check this in the new class tests.

File: tests/xharness/ITestReporter.cs [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+using Xharness.Listeners;
+using Xharness.Logging;
+
+namespace Xharness {
+
+	public interface ITestReporterFactory {
+		ITestReporter Create (IAppRunner appRunner, string device, ISimpleListener simpleListener, ILog log, ICrashSnapshotReporter crashReports);
+	}
+
+	// interface that represents a class that
```
Comment (wachaudh) [inline tests/xharness/ITestReporter.cs]: class that what?

File: tests/xharness/TestReporter.cs [hunk 17]
```diff
@@ -0,0 +1,490 @@
...
+				// MT1111: Application launched successfully, but it's not possible to wait for the app to exit as requested because it's not possible to detect app termination when launching using gdbserver
+				waitedForExit &= line?.Contains ("MT1111: ") != true;
+				if (line?.Contains ("error MT1007") == true)
```
Comment (wachaudh) [inline tests/xharness/TestReporter.cs]:
```suggestion
				if (line?.Contains ("error MT1007"))
```

File: tests/xharness/ITestReporter.cs (hunk 16 above)
Comment (mandel-macaque) [reply tests/xharness/ITestReporter.cs]: Agh, merge issue. will update.

File: tests/xharness/TestReporter.cs [hunk 18]
```diff
@@ -0,0 +1,490 @@
...
+	// about the success or failure of the execution.
+	public class TestReporter : ITestReporter {
+
+		const string timeoutMessage = "Test run timed out after {0} minute(s).";
+		const string completionMessage = "Test run completed";
+		const string failureMessage = "Test run failed";
+
+		public TimeSpan Timeout { get; private set; }
+		public IProcessManager ProcessManager { get; set; } = new ProcessManager ();
+		public IResultParser ResultParser { get; set; } = new XmlResultParser ();
```
Comment (premun) [inline tests/xharness/TestReporter.cs]: These should be private readonly and set from ctor

File: tests/xharness/AppRunner.cs [hunk 19]
```diff
@@ -432,31 +311,20 @@ public async Task<int> RunAsync ()
 
 			listener.StartAsync ();
 
-			var cancellation_source = new CancellationTokenSource ();
-			var timed_out = false;
+			// object that will take care of capturing and parsing the results
+			ILog run_log = MainLog;
+			var crashLogs = new Logs (Logs.Directory);
+			ICrashSnapshotReporter crashReporter = snapshotReporterFactory.Create (MainLog, crashLogs, isDevice: !isSimulator, deviceName);
+			var testResult = testReporterFactory.Create (this, deviceName, listener, run_log, crashReporter);
```
Comment (premun) [inline tests/xharness/AppRunner.cs]: >testResult [](start = 7, length = 10)

testResult is misleading

File: tests/xharness/TestReporter.cs [hunk 20]
```diff
@@ -0,0 +1,490 @@
...
+		readonly ICrashSnapshotReporter crashReporter;
+		readonly string deviceName;
+		bool waitedForExit = true;
+		bool launchFailure;
+		bool isSimulatorTest;
+		bool timedout;
+
+		readonly CancellationTokenSource cancellationTokenSource = new CancellationTokenSource ();
+
+		public TestReporter (IAppRunner appRunner, string device, ISimpleListener simpleListener, ILog log, ICrashSnapshotReporter crashReports)
```
Comment (premun) [inline tests/xharness/TestReporter.cs]: >IAppRunner appRunner [](start = 23, length = 20)

Let's not create dependency on AppRunner here as the TestReporter will go to the shared library. Just pass in the ILog and the RunMode or what is the reason you want this in.

File: tests/xharness/Xharness.Tests/Tests/AppRunnerTests.cs [hunk 21]
```diff
@@ -464,10 +477,20 @@ public async Task RunOnSimulatorSuccessfullyTest ()
...
-				.Setup (x => x.CleanXml (testResultFilePath, xmlResultFile))
-				.Callback (() => File.Copy (testResultFilePath, xmlResultFile));
+			var testReporterFactory = new Mock<ITestReporterFactory> ();
+			var testReporter = new Mock<ITestReporter> ();
+			testReporterFactory.Setup (f => f.Create (
+				It.IsAny<IAppRunner> (),
+				It.IsAny<string> (),
+				It.IsAny<ISimpleListener> (),
+				It.IsAny<ILog> (),
+				It.IsAny<ICrashSnapshotReporter> ())).Returns (testReporter.Object);
```
Comment (premun) [inline tests/xharness/Xharness.Tests/Tests/AppRunnerTests.cs]: It would be better if instead of `IsAny` you matched the values it should receive to be sure we for example don't send in wrong deviceName or something.

File: tests/xharness/AppRunner.cs [hunk 22]
```diff
@@ -25,15 +24,15 @@ class AppRunner
 		readonly ICrashSnapshotReporterFactory snapshotReporterFactory;
 		readonly ICaptureLogFactory captureLogFactory;
 		readonly IDeviceLogCapturerFactory deviceLogCapturerFactory;
-		readonly IResultParser resultParser;
+		readonly ITestReporterFactory testReporterFactory;
 
 		readonly RunMode runMode;
```
Comment (premun) [inline tests/xharness/AppRunner.cs]: I think `runMode` should stay private. But if you really need it public, remove this one and leave the public one only

**Comment by chamons** [comment]
Marked do not merge until xharness regressions are nailed down.

File: tests/xharness/TestReporter.cs (hunk 20 above)
Comment (mandel-macaque) [reply tests/xharness/TestReporter.cs]: AppRunner contains a number of details that we need, not only the loggers. And there are diff logers that are used. The two classes are coupled but via an interface, so we can have a diff Runner implementation, there is not need to use the one we already have. Which I still think we can reuse.

**Comment by mandel-macaque** [comment]
@chamons this PR fixes an issue were there are fake crash reports in vsts. removing the label, pushing to origin to get internal builds.

**Comment by mandel-macaque** [comment]
Ignore the 32 device not found errors, the pipeline was not properly configured. Failures in DDFun are unrelated and are due to tcp issues.
